File: rust/src/ixgbe/queues.rs
use crate::env::Environment;
use crate::lifed::{LifedArray, LifedPtr};

use super::buffer_pool::{Buffer, BufferPool};
use super::device::{Descriptor, Device, TransmitHead, RING_SIZE, RX_METADATA_DD, RX_METADATA_LENGTH, TX_METADATA_EOP, TX_METADATA_IFCS, TX_METADATA_LENGTH, TX_METADATA_RS};

pub struct QueueRx<'a> {
    ring: LifedArray<'a, Descriptor, { RING_SIZE }>,
    buffers: LifedArray<'a, LifedPtr<'a, Buffer<'a>>, { RING_SIZE }>,
    pool: LifedPtr<'a, BufferPool<'a>>,
    receive_tail_addr: LifedPtr<'a, u32>,
    next: u8,
}

impl<'a> QueueRx<'a> {
    pub fn create(env: &impl Environment<'a>, device: &Device<'a>, pool: LifedPtr<'a, BufferPool<'a>>) -> QueueRx<'a> {
        let ring = LifedArray::new(env.allocate::<Descriptor, { RING_SIZE }>());
        let buffers = LifedArray::new(env.allocate::<LifedPtr<'a, Buffer<'a>>, { RING_SIZE }>());

        for n in 0..RING_SIZE {
            let buffer = pool.map(|p| p.take()).unwrap(); // panic if the pool does not have enough buffers
            ring.index(n).write_volatile_part(buffer.read_volatile_part(|d| &d.phys_addr) as u64, |d| &mut d.addr);
            ring.index(n).write_volatile_part(0, |d| &mut d.metadata);
            buffers.set(n, buffer);
        }

        let receive_tail_addr = device.add_input(env, ring.index(0));
        receive_tail_addr.write_volatile(RING_SIZE as u32 - 1);

        QueueRx {
            ring,
            buffers,
            pool,
            receive_tail_addr,
            next: 0,
        }
    }

    #[inline(always)]
    pub fn batch(&mut self, buffers: &mut [LifedPtr<'a, Buffer<'a>>]) -> u8 {
        let mut rx_count: u8 = 0;
        while (rx_count as usize) < buffers.len() {
            let metadata = u64::from_le(self.ring.index(self.next as usize).read_volatile_part(|d| &d.metadata));
            if (metadata & RX_METADATA_DD) == 0 {
                break;
            }

            let new_buffer = match self.pool.map(|p| p.take()) {
                Some(b) => b,
                None => {
                    break;
                }
            };

            buffers[rx_count as usize] = self.buffers.get(self.next as usize);
            buffers[rx_count as usize].write_part(RX_METADATA_LENGTH(metadata), |b| &mut b.length);

            self.buffers.set(self.next as usize, new_buffer);
            self.ring.index(self.next as usize).write_volatile(Descriptor {
                addr: u64::to_le(new_buffer.map(|b| b.phys_addr) as u64),
                metadata: u64::to_le(0),
            });

            self.next = self.next.wrapping_add(1); // modulo RING_SIZE implicit since it's an u8
            rx_count = rx_count.wrapping_add(1);
        }
        if rx_count > 0 {
            // same, implicit modulo
            self.receive_tail_addr.write_volatile(u32::to_le(self.next.wrapping_sub(1) as u32));
        }
        rx_count
    }
}

const TX_RECYCLE_PERIOD: u8 = 32;

pub struct QueueTx<'a> {
    ring: LifedArray<'a, Descriptor, { RING_SIZE }>,
    buffers: LifedArray<'a, LifedPtr<'a, Buffer<'a>>, { RING_SIZE }>,
    pub pool: LifedPtr<'a, BufferPool<'a>>,
    transmit_head_addr: LifedPtr<'a, TransmitHead>,
    transmit_tail_addr: LifedPtr<'a, u32>,
    recycled_head: u8,
    next: u8,
}

impl<'a> QueueTx<'a> {
    pub fn create(env: &impl Environment<'a>, device: &Device<'a>, pool: LifedPtr<'a, BufferPool<'a>>) -> QueueTx<'a> {
        let ring = LifedArray::new(env.allocate::<Descriptor, { RING_SIZE }>());
        let transmit_head_addr = LifedPtr::new(&mut env.allocate::<TransmitHead, 1>()[0]);
        QueueTx {
            ring,
            buffers: LifedArray::new(env.allocate::<LifedPtr<'a, Buffer<'a>>, { RING_SIZE }>()),
            pool,
            transmit_head_addr,
            transmit_tail_addr: device.add_output(env, ring.index(0), transmit_head_addr),
            recycled_head: 0,
            next: 0,
        }
    }

    #[inline(always)]
    pub fn batch(&mut self, buffers: &mut [LifedPtr<'a, Buffer<'a>>]) -> u8 {
        if self.next.wrapping_sub(self.recycled_head) >= 2 * TX_RECYCLE_PERIOD {
            let actual_transmit_head = self.transmit_head_addr.read_volatile_part(|h| &h.value);
            while self.recycled_head as u32 != actual_transmit_head {
                if !self.pool.map(|p| p.give(self.buffers.get(self.recycled_head as usize))) {
                    break;
                }
                self.recycled_head = self.recycled_head.wrapping_add(1); // implicit modulo RING_SIZE, u8
            }
        }

        let mut tx_count: u8 = 0;
        while (tx_count as usize) < buffers.len() {
            if self.next == self.recycled_head.wrapping_sub(1) {
                break;
            }

            let rs_bit = if (self.next % TX_RECYCLE_PERIOD) == TX_RECYCLE_PERIOD - 1 { TX_METADATA_RS } else { 0 };
            self.ring.index(self.next as usize).write_volatile(Descriptor {
                addr: u64::to_le(buffers[tx_count as usize].map(|b| b.phys_addr) as u64),
                metadata: u64::to_le(TX_METADATA_LENGTH(buffers[tx_count as usize].map(|b| b.length)) | rs_bit | TX_METADATA_IFCS | TX_METADATA_EOP),
            });

            self.buffers.set(self.next as usize, buffers[tx_count as usize]);

            self.next = self.next.wrapping_add(1); // implicit modulo
            tx_count = tx_count.wrapping_add(1);
        }
        if tx_count > 0 {
            self.transmit_tail_addr.write_volatile(u32::to_le(self.next as u32));
        }
        tx_count
    }
}

File: csharp/TinyNF/Ixgbe/Queues.cs
using System;
using System.Runtime.CompilerServices;
using System.Threading;
using TinyNF.Environment;
using TinyNF.Unsafe;

namespace TinyNF.Ixgbe
{
    internal ref struct QueueRx
    {
        private readonly Array256<Descriptor> _ring;
        private readonly RefArray256<Buffer> _buffers;
        private readonly ref BufferPool _pool;
        private readonly ref uint _receiveTailAddr;
        private byte _next;

        public QueueRx(IEnvironment env, Device device, ref BufferPool pool)
        {
            _ring = new(env.Allocate<Descriptor>);
            _buffers = new(_ => ref Buffer.Fake);

            for (int n = 0; n < _ring.Length; n++)
            {
                ref var buffer = ref pool.Take(out bool valid);
                if (!valid)
                {
                    throw new Exception("Could not get a buffer to initialize the RX queue");
                }
                _ring[(byte)n].Addr = Endianness.ToLittle(buffer.PhysAddr);
                _ring[(byte)n].Metadata = Endianness.ToLittle(0);
                _buffers.Set((byte)n, ref buffer);
            }

            _pool = ref pool;
            _receiveTailAddr = ref device.SetInput(env, _ring.AsSpan()).Span[0];
            _next = 0;
        }

        // we have to use a refarray256 here otherwise using this without bounds check would be a huuuuge pain
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public byte Batch(RefArray256<Buffer> buffers, byte buffersCount)
        {
            byte rxCount = 0;
            while (rxCount < buffersCount)
            {
                ulong metadata = Endianness.FromLittle(Volatile.Read(ref _ring[_next].Metadata));
                if ((metadata & Device.RxMetadataDD) == 0)
                {
                    break;
                }

                ref var newBuffer = ref _pool.Take(out bool valid);
                if (!valid)
                {
                    break;
                }

                ref var returnedBuffer = ref _buffers.Get(_next);
                buffers.Set(rxCount, ref returnedBuffer);
                returnedBuffer.Length = Device.RxMetadataLength(metadata);

                _buffers.Set(_next, ref newBuffer);
                Volatile.Write(ref _ring[_next].Addr, Endianness.ToLittle(newBuffer.PhysAddr));
                Volatile.Write(ref _ring[_next].Metadata, Endianness.ToLittle(0));

                _next++; // implicit modulo since it's a byte
                rxCount++;
            }
            if (rxCount > 0)
            {
                Volatile.Write(ref _receiveTailAddr, Endianness.ToLittle((uint)(_next - 1)));
            }
            return rxCount;
        }
    }

    internal ref struct QueueTx
    {
        private const byte RecyclePeriod = 32;

        private readonly Array256<Descriptor> _ring;
        private readonly RefArray256<Buffer> _buffers;
        public readonly ref BufferPool Pool;
        private readonly ref TransmitHead _transmitHeadAddr;
        private readonly ref uint _transmitTailAddr;
        private byte _recycledHead;
        private byte _next;

        public QueueTx(IEnvironment env, Device device, ref BufferPool pool)
        {
            _ring = new(env.Allocate<Descriptor>);
            _buffers = new(_ => ref Buffer.Fake);
            Pool = ref pool;
            _transmitHeadAddr = ref env.Allocate<TransmitHead>(1).Span[0];
            _transmitTailAddr = ref device.AddOutput(env, _ring.AsSpan(), ref _transmitHeadAddr).Span[0];
            _recycledHead = 0;
            _next = 0;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public byte Batch(RefArray256<Buffer> buffers, byte buffersCount)
        {
            if ((byte)(_next - _recycledHead) >= 2 * RecyclePeriod)
            {
                uint actualTransmitHead = Endianness.FromLittle(Volatile.Read(ref _transmitHeadAddr.Value));
                while (_recycledHead != actualTransmitHead)
                {
                    if (!Pool.Give(ref _buffers.Get(_recycledHead)))
                    {
                        break;
                    }
                    _recycledHead++; // implicit modulo ring size since it's a byte
                }
            }

            byte txCount = 0;
            while (txCount < buffersCount)
            {
                if (_next == _recycledHead - 1)
                {
                    break;
                }

                ulong rsBit = _next % RecyclePeriod == RecyclePeriod - 1 ? Device.TxMetadataRS : 0;
                Volatile.Write(ref _ring[_next].Addr, Endianness.ToLittle(buffers.Get(txCount).PhysAddr));
                Volatile.Write(ref _ring[_next].Metadata, Endianness.ToLittle(Device.TxMetadataLength(buffers.Get(txCount).Length) | rsBit | Device.TxMetadataIFCS | Device.TxMetadataEOP));

                _buffers.Set(_next, ref buffers.Get(txCount));

                _next++; // implicit modulo
                txCount++;
            }
            if (txCount > 0)
            {
                Volatile.Write(ref _transmitTailAddr, Endianness.ToLittle(_next));
            }
            return txCount;
        }
    }
}
